File: python/mdat/src/mdat/core/formats/output/mdat.py
"""mdat per-frame TIFF export layout."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Callable

import numpy as np

from ..input import ImageInfo
from .base import ConvertSelection, ProgressCallback, emit_progress


class MdatOutputFormat:
    """Export one TIFF per (channel, time, z) frame under Pos*/ folders."""

    name = "mdat"

    def position_label(self, p_idx: int) -> str:
        return f"Pos{p_idx}"

    def run_convert(
        self,
        input_path: Path,
        output: Path,
        *,
        selection: ConvertSelection,
        info: ImageInfo,
        read_frame: Callable[[int, int, int, int], np.ndarray],
        on_progress: ProgressCallback | None,
    ) -> None:
        del input_path, info
        pos_indices = selection.pos_indices
        time_indices = selection.time_indices
        channel_indices = selection.channel_indices
        z_indices = selection.z_indices
        total = (
            len(pos_indices)
            * len(time_indices)
            * len(channel_indices)
            * len(z_indices)
        )
        emit_progress(
            on_progress,
            phase="start",
            done=0,
            total=total,
            message=(
                f"Selected {len(pos_indices)} positions, {len(time_indices)} timepoints, "
                f"{len(channel_indices)} channels, {len(z_indices)} z-slices. Total frames: {total}"
            ),
        )

        output.mkdir(parents=True, exist_ok=True)

        done = 0
        for p_idx in pos_indices:
            pos_dir = output / f"Pos{p_idx}"
            pos_dir.mkdir(exist_ok=True)

            with open(pos_dir / "time_map.csv", "w", newline="", encoding="utf-8") as fh:
                writer = csv.writer(fh)
                writer.writerow(["t", "t_real"])
                for t_new, t_orig in enumerate(time_indices):
                    writer.writerow([t_new, t_orig])

            for t_new, t_orig in enumerate(time_indices):
                for c_orig in channel_indices:
                    for z_orig in z_indices:
                        frame = read_frame(p_idx, t_orig, c_orig, z_orig)
                        filename = (
                            f"img_channel{c_orig:03d}"
                            f"_position{p_idx:03d}"
                            f"_time{t_new:09d}"
                            f"_z{z_orig:03d}.tif"
                        )
                        write_tiff(pos_dir / filename, frame)
                        done += 1
                        emit_progress(
                            on_progress,
                            phase="advance",
                            done=done,
                            total=total,
                            message="Writing TIFFs",
                        )

        emit_progress(
            on_progress,
            phase="finish",
            done=done,
            total=total,
            message=f"Wrote {output}",
        )


def write_tiff(path: Path, frame: np.ndarray) -> None:
    """Write a TIFF, replacing an existing file first for robust reruns on Windows."""
    import tifffile

    if path.exists():
        path.unlink()
    tifffile.imwrite(str(path), frame)


__all__ = ["MdatOutputFormat", "write_tiff"]

File: crates/mdat/src/output/mdat.rs
use std::path::Path;

use crate::error::Result;
use crate::input::ImageInfo;
use crate::input::ReaderSession;
use crate::io::write_tiff;
use crate::output::{emit_progress, OutputFormatWriter, ProgressPhase};
use crate::selection::ConvertSelection;

pub struct MdatOutputFormat;

impl OutputFormatWriter for MdatOutputFormat {
    fn name(&self) -> &'static str {
        "mdat"
    }

    fn position_label(&self, p_idx: usize) -> String {
        format!("Pos{p_idx}")
    }

    fn run_convert(
        &self,
        _input_path: &Path,
        output: &Path,
        selection: &ConvertSelection,
        _info: &ImageInfo,
        session: &mut ReaderSession,
        on_progress: &mut Option<&mut dyn crate::output::ProgressCallback>,
    ) -> Result<()> {
        let total = selection.pos_indices.len()
            * selection.time_indices.len()
            * selection.channel_indices.len()
            * selection.z_indices.len();

        emit_progress(
            on_progress,
            ProgressPhase::Start,
            0,
            total,
            format!(
                "Selected {} positions, {} timepoints, {} channels, {} z-slices. Total frames: {total}",
                selection.pos_indices.len(),
                selection.time_indices.len(),
                selection.channel_indices.len(),
                selection.z_indices.len(),
            ),
        );

        std::fs::create_dir_all(output)?;
        let mut done = 0usize;

        for &p_idx in &selection.pos_indices {
            let pos_dir = output.join(format!("Pos{p_idx}"));
            std::fs::create_dir_all(&pos_dir)?;

            // Use original source time indices in filenames (gaps allowed).
            // Downstream tools (e.g. lisca folder source) discover times from
            // names and support non-contiguous series — no time_map renumbering.
            for &t_orig in &selection.time_indices {
                for &c_orig in &selection.channel_indices {
                    for &z_orig in &selection.z_indices {
                        let frame = session.read_frame(p_idx, t_orig, c_orig, z_orig)?;
                        let filename = format!(
                            "img_channel{c_orig:03}_position{p_idx:03}_time{t_orig:09}_z{z_orig:03}.tif"
                        );
                        write_tiff(
                            &pos_dir.join(filename),
                            &frame,
                            session.width,
                            session.height,
                        )?;
                        done += 1;
                        emit_progress(
                            on_progress,
                            ProgressPhase::Advance,
                            done,
                            total,
                            "Writing TIFFs",
                        );
                    }
                }
            }
        }

        emit_progress(
            on_progress,
            ProgressPhase::Finish,
            done,
            total,
            format!("Wrote {}", output.display()),
        );
        Ok(())
    }
}
